x-pack/plugins/rule_registry/server/rule_data_plugin_service/index_names.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Dataset } from './index_options';
+import { ResourceNames } from './resource_names';
+
+const joinWithDash = ResourceNames.joinWithDash;
+
+interface ConstructorOptions {
+  resourceNames: ResourceNames;
+
+  /** @example 'security', 'observability', 'observability.logs' */
+  registrationContext: string;
+
+  /** @example 'alerts', 'events' */
+  dataset: Dataset;
+
+  /** @example '.siem-signals', null */
+  secondaryAlias: string | null;
+}
+
+export class IndexNames {
+  constructor(private readonly options: ConstructorOptions) {
+    const { resourceNames, registrationContext, dataset } = options;
+
+    this.prefix = resourceNames.getFullPrefix();
+    this.baseName = resourceNames.getFullName(registrationContext, dataset);
+    this.basePattern = joinWithDash(this.baseName, '*');
+  }
+
+  /** @example '.alerts' */
+  public readonly prefix: string;
+
+  /** @example '.alerts-security.alerts' */
+  public readonly baseName: string;
+
+  /** @example '.alerts-security.alerts-*' */
+  public readonly basePattern: string;
+
+  /** @example '.alerts-security.alerts-default' */
+  public getPrimaryAlias(namespace: string): string {
+    return joinWithDash(this.baseName, namespace);
+  }
+
+  /** @example '.alerts-security.alerts-default-*' */
+  public getPrimaryAliasPattern(namespace: string): string {
```

Comment:
Not sure I understood your idea @marshallmain, why is it dangerous?

I will probably add one or two additional functions here for backing indices, and maybe rename this one to better indicate that this is _not_ related to backing indices.

Reply:
2 reasons that it's dangerous: the "namespaces with dashes" problem where `-*` could unintentionally match multiple namespaces, and also it encourages bypassing the alias to query all of the backing indices at once.

We should never need `.alerts-security.alerts-default-*` as a general pattern except when dealing with backing indices though right? And for the backing indices we would want `.internal-.alerts-security.alerts-default-*` eventually as the pattern for a particular namespace, and if we want to query all namespaces for an alias we'd use `.alerts-security.alerts-*`.

Reply:
As I understand @marshallmain's argument, once we introduce the the `.internal` prefix separation, there is no valid use-case for the value returned by this function because it should never match anything. Until then it's potentially dangerous because it targets both aliases and backing indices.

Reply:
> We should never need .alerts-security.alerts-default-* as a general pattern except when dealing with backing indices though right?

After we introduce them, yep. At the moment `getPrimaryAliasPattern()` is used in `index_patterns` of the index template and in `createWriteTargetIfNeeded`.

> And for the backing indices we would want .internal-.alerts-security.alerts-default-* eventually as the pattern for a particular namespace

Yep

> and if we want to query all namespaces for an alias we'd use .alerts-security.alerts-*.

Yep

Leaving this function as is in this PR.